In [1]:
from embedder import Embedder
from data_provider import DataProvider

In [2]:
import os
os.environ['TRANSFORMERS_CACHE'] = '/mounts/work/faeze/.cache/hf/'
os.environ['HF_HOME'] = '/mounts/work/faeze/.cache/hf/'
os.environ['HF_DATASETS_CACHE'] = '/mounts/work/faeze/.cache/hf/'
os.environ['TORCH_HUB'] = '/mounts/work/faeze/.cache/torch/'
os.environ['TORCH_HOME'] = '/mounts/work/faeze/.cache/torch/'
os.environ["WANDB_DIR"] = '/mounts/work/faeze/.cache/wandb/'

In [3]:
data_provider_ = DataProvider()

In [4]:
languages = ['en', 'tr', 'de', 'hr', 'ru', 'sq']
types = ['Gao', 'Trac', 'Wul']

datas = [f'parallel_data/{i}-Trac' for i in languages]

In [5]:
datasets = data_provider_.load_datasets(datas, languages)

/mounts/work/faeze/new_miniconda3/envs/2024_9/lib/python3.10/site-packages/datasets/load.py:929: FutureWarning: The repository for parallel_data contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at parallel_data/parallel_data.py
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


In [6]:
datasets

[{'name': 'parallel_data/en-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'en'},
 {'name': 'parallel_data/tr-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'tr'},
 {'name': 'parallel_data/de-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'de'},
 {'name': 'parallel_data/hr-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'hr'},
 {'name': 'parallel_data/ru-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'ru'},
 {'name': 'parallel_data/sq-Trac',
  'split': 'test',
  'data': Dataset({
      features: ['id', 'text', 'label'],
      num_rows: 300
  }),
  'language': 'sq'}]

In [7]:
def get_similarity(model_name):
    embedder_ = Embedder(model_name=model_name)
    embeddings, metadatas = embedder_.embed_datasets(datasets)
    
    similarities = [[Embedder.calculate_similarity(embeddings[i], embeddings[j])[1] for i in range(len(embeddings))] for j in range(len(embeddings))]
    
    return similarities

In [8]:
languages = ['en', 'tr', 'de', 'hr', 'ru', 'sq']

In [9]:
get_similarity('labse')

/mounts/Users/cisintern/faeze/.local/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


[[0.9158988, 0.753104, 0.7741467, 0.7709965, 0.75221217, 0.76883596],
 [0.7537379, 0.86096406, 0.72797996, 0.72831756, 0.7112424, 0.7246551],
 [0.77429724, 0.7272874, 0.85106, 0.7517183, 0.7309431, 0.7422168],
 [0.7714623, 0.72731763, 0.7513352, 0.8776765, 0.7345994, 0.7459766],
 [0.7511185, 0.7092914, 0.729755, 0.73415625, 0.8545852, 0.71943617],
 [0.769398, 0.7249907, 0.743189, 0.7469721, 0.72135884, 0.8566743]]

In [10]:
get_similarity('minilm')

[[0.9116931, 0.72609895, 0.76991445, 0.7451464, 0.74737597, 0.7518816],
 [0.72787493, 0.8952419, 0.7420276, 0.7293082, 0.7176411, 0.72639614],
 [0.77237535, 0.7426334, 0.89149094, 0.7623501, 0.74733025, 0.7535743],
 [0.74779993, 0.73032093, 0.762792, 0.89622086, 0.75045806, 0.75532836],
 [0.7484968, 0.7180144, 0.7471681, 0.74964166, 0.8983518, 0.730639],
 [0.7535543, 0.72655106, 0.75324607, 0.75446236, 0.7304253, 0.9006354]]

In [11]:
get_similarity('e5-large')

[[0.59967417, 0.5232911, 0.532949, 0.5292311, 0.5262704, 0.5290959],
 [0.52296937, 0.59555864, 0.51909304, 0.51708394, 0.51535285, 0.5197444],
 [0.5329468, 0.519225, 0.59310615, 0.5223896, 0.5213238, 0.52084565],
 [0.5296454, 0.517395, 0.52236885, 0.59423864, 0.5238726, 0.5378648],
 [0.52621305, 0.5154257, 0.52082276, 0.52343976, 0.60073006, 0.52866447],
 [0.5291786, 0.519864, 0.5203655, 0.5372616, 0.5283685, 0.6027343]]

In [12]:
get_similarity('e5-base')

[[0.5984787, 0.514099, 0.52812517, 0.51835144, 0.517258, 0.5209366],
 [0.5140816, 0.5946191, 0.5126872, 0.51127774, 0.50833505, 0.5139458],
 [0.5280132, 0.51257515, 0.59377646, 0.51529217, 0.5109679, 0.51471853],
 [0.5187787, 0.5114007, 0.5153993, 0.5976846, 0.5171343, 0.53263843],
 [0.51717424, 0.508206, 0.51079226, 0.5169573, 0.59986854, 0.5202148],
 [0.5210433, 0.5138775, 0.5146746, 0.5321724, 0.51990235, 0.60140026]]

In [13]:
get_similarity('xlmr-large')

No sentence-transformers model found with name FacebookAI/xlm-roberta-large. Creating a new one with mean pooling.


[[0.49967372, 0.49698266, 0.49737033, 0.49684817, 0.4974424, 0.49597836],
 [0.4970168, 0.4996596, 0.49668834, 0.49641448, 0.4968395, 0.49564227],
 [0.49738622, 0.49667546, 0.4995246, 0.49649438, 0.49717203, 0.49567232],
 [0.49690828, 0.4964378, 0.49653605, 0.49960858, 0.49681267, 0.49576437],
 [0.49744704, 0.4968036, 0.49715668, 0.49677473, 0.49934793, 0.49584454],
 [0.49608633, 0.49570617, 0.49576104, 0.49580613, 0.49592388, 0.49977827]]

In [14]:
get_similarity('xlmr-base')

No sentence-transformers model found with name FacebookAI/xlm-roberta-base. Creating a new one with mean pooling.


[[0.49999857, 0.4974318, 0.4979207, 0.49773958, 0.49792868, 0.49648324],
 [0.49751103, 0.4998691, 0.49738288, 0.4973874, 0.49737722, 0.4965954],
 [0.49796894, 0.49733698, 0.49982315, 0.49739942, 0.49779734, 0.49662665],
 [0.4978263, 0.49737707, 0.49742863, 0.49975783, 0.49770036, 0.49667838],
 [0.49795341, 0.49732783, 0.4978162, 0.49767384, 0.49971825, 0.49659175],
 [0.49670836, 0.49669203, 0.49676937, 0.49682096, 0.4966995, 0.499888]]

In [15]:
get_similarity('arctic-large')

[[0.85811996, 0.634602, 0.72069883, 0.64986295, 0.6523283, 0.64447737],
 [0.6341692, 0.8496975, 0.62637943, 0.6076001, 0.5977409, 0.6101656],
 [0.7201146, 0.6260352, 0.8679098, 0.65407264, 0.6515242, 0.64619136],
 [0.65028316, 0.6080188, 0.65451694, 0.8432339, 0.63529485, 0.63326216],
 [0.65066427, 0.59612626, 0.65053725, 0.6333518, 0.86878353, 0.6134701],
 [0.64417565, 0.6100859, 0.64611053, 0.6328484, 0.6141701, 0.86161494]]

In [16]:
get_similarity('arctic-base')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/252 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/85.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/107 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/738 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.38k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/712k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/695 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

[[0.6573355, 0.40095198, 0.4562814, 0.40605754, 0.34360576, 0.41121536],
 [0.39821965, 0.5967858, 0.44899452, 0.4439353, 0.4260963, 0.44909748],
 [0.45400053, 0.44973218, 0.61287177, 0.44606912, 0.40361145, 0.4530835],
 [0.4024778, 0.44319302, 0.44464004, 0.61208534, 0.40036598, 0.45000532],
 [0.3394619, 0.42798293, 0.40331328, 0.40183234, 0.54402506, 0.40494958],
 [0.4083741, 0.44882563, 0.45231932, 0.45088458, 0.4034228, 0.61086076]]

In [17]:
get_similarity('m3')

[[0.7891197, 0.6449263, 0.6807311, 0.66090626, 0.65425205, 0.65964854],
 [0.64412737, 0.79690355, 0.6316072, 0.626419, 0.61489666, 0.62981987],
 [0.68036616, 0.6317699, 0.80116624, 0.65442884, 0.63596064, 0.647951],
 [0.66144466, 0.6275461, 0.6552952, 0.8031053, 0.6435715, 0.6512648],
 [0.6538418, 0.61541057, 0.6362567, 0.64273715, 0.8017712, 0.6293741],
 [0.65950584, 0.63011587, 0.6482431, 0.6502399, 0.62924826, 0.80063635]]

In [19]:
get_similarity('m3-unsupervised')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/123 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/13.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/54.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/687 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.31k [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

[[0.7831914, 0.6402351, 0.6707782, 0.6546389, 0.6527621, 0.6558595],
 [0.6404313, 0.7775309, 0.62592685, 0.6221998, 0.61593467, 0.62516755],
 [0.67130363, 0.6261263, 0.7780322, 0.64642775, 0.63282615, 0.64061713],
 [0.6549732, 0.6221683, 0.6461831, 0.7830085, 0.6386269, 0.6436403],
 [0.65288925, 0.6163039, 0.6329517, 0.6389356, 0.7849221, 0.62869996],
 [0.6558198, 0.62481695, 0.6401061, 0.6431704, 0.627947, 0.7828696]]

In [20]:
get_similarity('m3-retromae')

No sentence-transformers model found with name BAAI/bge-m3-retromae. Creating a new one with mean pooling.


config.json:   0%|          | 0.00/694 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.16G [00:00<?, ?B/s]

Some weights of XLMRobertaModel were not initialized from the model checkpoint at BAAI/bge-m3-retromae and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/419 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

[[0.5481223, 0.47015032, 0.48061907, 0.46748677, 0.47161585, 0.46524405],
 [0.47037706, 0.55090916, 0.46515894, 0.45892105, 0.46043685, 0.45687467],
 [0.48095253, 0.4651985, 0.5470413, 0.463034, 0.46503067, 0.45931464],
 [0.4678292, 0.45920864, 0.46310726, 0.54976594, 0.46386895, 0.45814413],
 [0.472395, 0.46067807, 0.4653046, 0.4643758, 0.5457804, 0.45475864],
 [0.46511865, 0.45669636, 0.45885727, 0.4577304, 0.45399514, 0.5403682]]

In [48]:

embedder_ = Embedder(model_name="m3")
embeddings, metadatas = embedder_.embed_datasets(datasets)

In [49]:
import numpy as np
embed = np.vstack(embeddings)
metadata = metadatas

In [50]:
len(metadata), embed.shape

(1800, (1800, 1024))

In [51]:
embed.shape

(1800, 1024)

In [52]:
clusters = Embedder.cluster_embeddings(embed, metadata, 150)
clusters

text  label   id  \
0     Even 2 month old baby girl is not safe in Indi...      0    0   
1     these bhagwa terrorists can't digest their mea...      1    1   
2     What kind of media is this!!first they make he...      0    2   
3     Compensation to the States will be payable eve...      0    3   
4     Nifty below 8550 , indicates a down move maybe...      0    4   
...                                                 ...    ...  ...   
1795  ish kryeministri I madh I Indisë hera e parë f...      0  295   
1796  modaliteti Mute i çaktivizuar? Epo të paktën d...      1  296   
1797  kryeministër I dështuar, skemë pa para kesh e ...      1  297   
1798  Propaganda kundër Pawar Saab është krejtësisht...      0  298   
1799  India ka nevojë për një kryeministër të tillë ...      0  299   

               dataset_name language  cluster  
0     parallel_data/en-Trac       en      124  
1     parallel_data/en-Trac       en       83  
2     parallel_data/en-Trac       en       30  
3     parallel_data/en-Trac       en      126  
4     parallel_data/en-Trac       en       60  
...                     ...      ...      ...  
1795  parallel_data/sq-Trac       sq       76  
1796  parallel_data/sq-Trac       sq       33  
1797  parallel_data/sq-Trac       sq       66  
1798  parallel_data/sq-Trac       sq       29  
1799  parallel_data/sq-Trac       sq       76  

[1800 rows x 6 columns]

In [53]:
clusters[clusters['cluster']==8]

text  label   id  \
14    No quotas no reservations! First the UN Should...      1   14   
144   #Niaz_Khosa google it in 2017 u r 6th most rap...      1  144   
190   And who do you think YOU are, to decide, who's...      1  190   
314   Kota yok rezervasyon yok !İlk olarak BM kadını...      1   14   
444   #Niaz_Khosa 2017 için google’dan arat siz  6. ...      1  144   
490   Ve SEN kim olduğunu düşünüyorsun da kimin içer...      1  190   
614   Keine Quoten keine Vorbehalte! Zuerst sollte d...      1   14   
744   #Niaz_Khosa google es in 2017 und du bist das ...      1  144   
790   Und was glaubst du, wer DU bist, um zu entsche...      1  190   
914   Nema kvota nema rezervacija! Prvo bi UN trebao...      1   14   
1044  #Niaz_Khosa Guglaj to, u 2017. vi ste 6. po re...      1  144   
1090  A što misliš tko si ti da odlučiš tko je insaj...      1  190   
1214  Нет квоты нет оговорок! Сначала ООН должна улу...      1   14   
1344  #Niaz_Khosa загугли это в 2017 году вы были на...      1  144   
1390  "И кто ты, по-твоему, такой, чтобы решать, кто...      1  190   
1514  No Kuotat nuk ka rezerva! Së pari OKB-ja duhet...      1   14   
1644  #Niaz_Khosa kërkoje në google në vitin 2017 je...      1  144   
1690  Dhe ju, kush mendoni se jeni, për të vendosur,...      1  190   

               dataset_name language  cluster  
14    parallel_data/en-Trac       en        8  
144   parallel_data/en-Trac       en        8  
190   parallel_data/en-Trac       en        8  
314   parallel_data/tr-Trac       tr        8  
444   parallel_data/tr-Trac       tr        8  
490   parallel_data/tr-Trac       tr        8  
614   parallel_data/de-Trac       de        8  
744   parallel_data/de-Trac       de        8  
790   parallel_data/de-Trac       de        8  
914   parallel_data/hr-Trac       hr        8  
1044  parallel_data/hr-Trac       hr        8  
1090  parallel_data/hr-Trac       hr        8  
1214  parallel_data/ru-Trac       ru        8  
1344  parallel_data/ru-Trac       ru        8  
1390  parallel_data/ru-Trac       ru        8  
1514  parallel_data/sq-Trac       sq        8  
1644  parallel_data/sq-Trac       sq        8  
1690  parallel_data/sq-Trac       sq        8

In [54]:
embed.shape, len(metadata)

((1800, 1024), 1800)

In [55]:
Embedder.visualize_embeddings(embed, metadatas)

/mounts/work/faeze/new_miniconda3/envs/2024_9/lib/python3.10/site-packages/umap/umap_.py:1952: UserWarning:

n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.



In [56]:
from retriever import Retriever
retriever = Retriever(embedder_.embedding_dim, index_type="FlatL2")

In [57]:
retriever.add_embeddings(embed, metadata)

In [58]:
retriever.retrieve(embeddings[0][213].reshape(1,-1), k=10)

[{'metadata': {'text': "Stop fooling people with your Jumlas. Address to the issues like Employment Education Poverty Health Farmers Infrastructure & Economic Reforms. Please don't divert people's attention from core issues which matters to the nation. Nation wants concrete result not Jumlas.",
   'label': 1,
   'id': '213',
   'dataset_name': 'parallel_data/en-Trac',
   'language': 'en'},
  'score': 0.0},
 {'metadata': {'text': 'Hört auf die Leuten mit euren Jumlas zu verarschen. Sprecht die Themen wie Beschäftigung, Bildung, Armut, Gesundheit, Bauern, Infrastruktur & Wirtschaftliche Reformen an. Bitte lenkt die Aufmerksamkeit der Menschen nicht von den Kernfragen ab, die für die Nation wichtig sind. Das Land will konkrete Ergebnisse, nicht Jumlas.',
   'label': 1,
   'id': '213',
   'dataset_name': 'parallel_data/de-Trac',
   'language': 'de'},
  'score': 0.11470854},
 {'metadata': {'text': 'Хватит обманывать людей своими джамлами. Обращайтесь к таким вопросам, как образование в обла

In [59]:
metadata[213]

{'text': "Stop fooling people with your Jumlas. Address to the issues like Employment Education Poverty Health Farmers Infrastructure & Economic Reforms. Please don't divert people's attention from core issues which matters to the nation. Nation wants concrete result not Jumlas.",
 'label': 1,
 'id': '213',
 'dataset_name': 'parallel_data/en-Trac',
 'language': 'en'}

In [60]:
ks = [10, 20, 30, 40, 50]
offset = 1 + 5

averages = []
for k in ks:
    counts = []
    for i in range(300):
        retrieved = retriever.retrieve(embeddings[0][i].reshape(1,-1), k=k+offset)
        # print(metadata[i])
        # print(retrieved[6:])
        lang_en_ids = [item["metadata"]["id"] for item in retrieved[offset:] if item["metadata"]["language"] == "en"]
        
        # count = sum(1 for item in retrieved[6:] if item["metadata"]["id"] in lang_en_ids)
        count = sum(1 for item in retrieved[offset:] if item["metadata"]["id"] in lang_en_ids and item["metadata"]["language"] != "en")
        counts.append(count/k)
    averages.append(np.round(np.mean(counts),3))
averages   

[0.468, 0.516, 0.54, 0.55, 0.559]


"e5-large": [0.164, 0.2165, 0.247, 0.2705, 0.292]

"e5-base": [0.153, 0.212, 0.246, 0.27, 0.288]

"labse": [0.376, 0.428, 0.438, 0.446, 0.448]

"minilm": [0.092, 0.132, 0.134, 0.143, 0.157]

"xlmr-large": [0.001, 0.001, 0.001, 0.002, 0.002]

"m3" : [0.47, 0.516, 0.54, 0.548, 0.553]

"m3-unsupervised" : [0.441, 0.49, 0.515, 0.523, 0.525]

"arctic-large": [0.414, 0.476, 0.508, 0.523, 0.532]

"arctic-base": [0.026, 0.049, 0.072, 0.09, 0.108]





